### DISCUSSION TITLE
Argument syntax

### DISCUSSION THREAD

**Discussion by adesutherland** [adesutherland/CREXX#126] [Ideas]
I have been thinking about the way to define arguments / procedure in REXX level b. 

Needs to be type safe and also needs to be REXX friendly; I note feedback that having or not having a "=" to represent pass by value vs pass by reference would be error prone.

So I am thinking of this:

```
test: procedure = .int
  arg a = .int, expose b = .float, c = 0.0, d. = .int
```
In this example
* test is a procedure that returns an int
* a is a int (pass by value - or copy)
* b is a float, it is exposed so it is "obviously" passed by reference - or move
* c is an optional float with a default value of 0.0
* d is an array of integers  (see below for some syntax candy)

So we can call it like
```
x = 1.2
y. = .int
y.1 = 20
y.2 = 30
call test 5, x, , y
```
We can also call it like this (showing the syntax candy to have an undefined number of arguments; 20 and 30 are made into array d)
```
x = 1.2
call test 5, x, , 20, 30
```

And finally some expose syntax candy for making it more "REXX"
```
exp = 100

say test( "hello")
exit

test: procedure = .string expose exp = .int
  arg message = .string
  return message || exp
```

The compiler converts this to
```
exp = 100

say test(exp, "hello")
exit

test: procedure = .string 
  arg expose exp = .int, message = .string
  return message || exp
```

This is designed to provide a familiar (but not the same) experience for REXX programmers

Comments welcome!

**Comment by Peter-Jacob** [comment]
my response time is slow, we are all day on tour, the is not much room for checking the internet. We are back on Thursday then it will be better!
```
x = 1.2
call test 5, x, , 20, 30
```
It looks nice but may have one problem if you use more than one stem alike variable, how do you distinguish them?

```
x = 1.2
y. = .int
y.1 = 20
y.2 = 30
z. = .int
z.1 = 100
z.2 = 101
call test 5, x, , y,z
```
in this case, the shortened form wouldn't work properly

**Comment by adesutherland** [comment]
You are right - the candy will have to only work for the last "stem" variable! Luckily I don't need to worry about it yet as cREXX does not yet support these arrays :-)

Enjoy your tour back - say hello to Zurich and the black forest if you go that way, or have some decent food in the Alsace for me - I am stuck here in the UK ... :-(
